rust/core-lib/src/plugins/manifest.rs

```diff
@@ -107,10 +111,131 @@ pub enum PluginScope {
     SingleInvocation,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+/// Represents a custom command provided by a plugin.
+pub struct Command {
+    /// Human readable title, for display in (for example) a menu.
+    pub title: String,
+    /// A short description of the command.
+    pub description: String,
+    /// Template of the command RPC as it should be sent to the plugin.
+    pub rpc_cmd: PlaceholderRpc,
+    /// A list of `CommandArgument`s, which the client should use to build the RPC.
+    pub args: Vec<CommandArgument>,
+}
+
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+/// A user provided argument to a plugin command.
+pub struct CommandArgument {
+    /// A human readable name for this argument, for use as placeholder
+    /// text or equivelant.
+    pub title: String,
+    /// A short (single sentence) description of this argument's use.
+    pub description: String,
+    pub key: String,
+    pub arg_type: ArgumentType,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(tag = "type_name")]
```

Comment:
I gave this some thought, and `type_name` was the most semantically communicative key I could come up with. In my experience, `id` is mostly used as a unique identifier (as in the json-rpc protocol, or scads of REST APIs).

My inclination is to revert this change and go back to having `arg_type`/`type` be a string, and then have the optional `options` field. Using serde's 'either string or struct' doesn't offer any more safety than an assert, and it potentially complicates parsing for clients.

Keep in mind that these types are never actually used in in rust; they're deserialized in one format and then serialized in another.